src/operation.rs

```diff
@@ -69,14 +71,271 @@ impl Operation {
+/// 
+/// - `operations` corresponds to the sequence of operations
+/// - `measurement_targets` corresponds to the targets of the measurement operations
+/// - `conditional_operations` corresponds to the sequence of conditional operations
+/// 
+pub struct QuantumCircuit {
+    pub(crate) operations: Vec<Operation>,
+    measurement_targets: Vec<usize>,
+    conditional_operations: Vec<(Operation, Vec<(usize, bool)>)>,
+}
+
+impl QuantumCircuit {
+    /// Constructs a new empty quantum circuit.
+    pub fn new() -> QuantumCircuit {
+        QuantumCircuit {
+            operations: Vec::new(),
+            measurement_targets: Vec::new(),
+            conditional_operations: Vec::new(),
+        }
+    }
+    /// Clears the operations in the quantum circuit.
+    pub fn reset_register(&mut self) -> &mut Self {
+        self.operations.clear();
+        self.measurement_targets.clear();
+        self.conditional_operations.clear();
+        self
+    }
+    pub fn get_operations(&self) -> &Vec<Operation> {
+        &self.operations
+    }
+    pub fn get_measurement_targets(&mut self) -> &Vec<usize> {
+        &self.measurement_targets
+    }
+    pub fn get_conditional_operations(&mut self) -> &Vec<(Operation, Vec<(usize, bool)>)> {
+        &self.conditional_operations
+    }
+
+    /// Reduces the circuit by combining gates with overlapping targets.
+    /// Migth be slower if used after 
+    /// [`reduce_non_overlapping_gates`](#method.reduce_non_overlapping_gates)
+    pub fn reduce_gates_with_one_off_size(&mut self, max_size : usize) -> &mut Self {
+        //No gates of same size on same targets left
+        self.reduce_circuit_gates_with_same_targets();
+        let mut i = 0;
+        let ops = &self.operations;
+        let mut new_ops: Vec<Operation> = Vec::new();
+        let len = ops.len(); 
+        while i < len {
+            if i == len - 1 {
+                new_ops.push(ops[i].clone());
+                break;
+            }
+            let j = i + 1;
+            let op = &ops[i];
+            let next_op = &ops[j];
+            if op.targets().len() > max_size || next_op.targets().len() > max_size {
+                new_ops.push(op.clone());
+                i += 1;
+                continue;
+            }
+            if self.operations_overlap(op.clone(), next_op.clone()){
+                let combined_op = self.combine_ops(op.clone(), next_op.clone());
+                new_ops.push(combined_op);
+                i += 2;
+            } else {
+                new_ops.push(op.clone());
+                i += 1;
+            }
+
+        }
+        self.operations = new_ops;
+        self
+    }
+    /// Combines two operations with overlapping targets
+    fn combine_ops(&self, op1 : Operation, op2 : Operation) -> Operation {
+        if op1.targets().len() > 2 || op2.targets().len() > 2 {
+            panic!("Cannot combine operations with more than 2 targets");
+        }
+        let mut matrix = op1.matrix();
```

Comment:
Change
```
        let mut matrix = op1.matrix();
        let mut targets = op1.targets().clone();
        let diff : i32 = op1.targets.len() as i32- op2.targets().len() as i32;
        //op1 is smaller
        if diff < 0 {
            //extend bottom
            targets = op2.targets().clone();
            if op1.targets[0] == op2.targets()[0]{
                matrix = linalg::kron(&Array2::eye(2), &op1.matrix());
                matrix = op2.matrix().dot(&matrix);
            }
            else {
                matrix = linalg::kron(&op1.matrix(), &Array2::eye(2));
                matrix = op2.matrix().dot(&matrix);
            }
        }
        else {
            targets = op1.targets().clone();
            if op2.targets[0] == op1.targets[0] {
                matrix = linalg::kron(&Array2::eye(2), &op2.matrix());
                matrix = op1.matrix().dot(&matrix);
            }
            else {
                matrix = linalg::kron(&op2.matrix(), &Array2::eye(2));
                matrix = op1.matrix().dot(&matrix);
            }
        }
```
to (for example)
```
        let diff: i32 = op1.targets.len() as i32 - op2.targets().len() as i32;

        let targets = if diff < 0 {
            op2.targets().clone()
        } else {
            op1.targets().clone()
        };

        let matrix = if diff < 0 {
            // op1 is smaller, extend bottom
            if op1.targets[0] == op2.targets()[0] {
                op2.matrix()
                    .dot(&linalg::kron(&Array2::eye(2), &op1.matrix()))
            } else {
                op2.matrix()
                    .dot(&linalg::kron(&op1.matrix(), &Array2::eye(2)))
            }
        } else {
            if op2.targets[0] == op1.targets[0] {
                op1.matrix()
                    .dot(&linalg::kron(&Array2::eye(2), &op2.matrix()))
            } else {
                op1.matrix()
                    .dot(&linalg::kron(&op2.matrix(), &Array2::eye(2)))
            }
        };
```
or something else that fixes the `unused value` warning